### PR TITLE
Detect heap allocation to constrain Allocation mode

### PR DESCRIPTION
**Pull request by zhizhuoxin**
- Detect all kinds of possible heap allocation to determine mode for the Allocation mode axis
- Distinguish stack allocations marked by `stack_` from others - these stack allocations do not force the closure's mode to be alloc
- Handle corner cases conservatively: (1) `Prim_poly` primitives; (2) multi-argument (curried) function definitions; (3) empty/non-capturing functors

Testing
-------
Add more tests in `testsuite/tests/typing-modes/zero_alloc.ml`
- If a closure contains any kind of allocation, the closure is forced to be alloc
- Allocation mode axis check skips stack allocations (marked by `stack_`)
- Test several special cases that are handled conservatively

### CODE REVIEW

File: typing/env.ml
Comment (dkalinichenko-js) [inline typing/env.ml:1]: Since https://github.com/oxcaml/oxcaml/pull/6230 is merged now, let's rebase onto main so the changes from there don't show up in the diff.

File: typing/env.ml [hunk 1]
```diff
@@ -3725,30 +3721,48 @@ let unboxed_type ~errors ~env ~loc ~lid ty =
...
       | Exclave_lock ->
-          exclave_mode ~errors ~env ~loc ~item ~lid vmode
+          exclave_mode ~errors ~env ~pp vmode
       | Unboxed_lock ->
-          unboxed_type ~errors ~env ~loc ~lid ty;
+          unboxed_type ~errors ~env ~loc:(fst pp) ty_and_lid;
           vmode
     ) mode locks
 
+(** Registers a use of an allocation, constraining every enclosing closure lock.
```
Comment (dkalinichenko-js) [inline typing/env.ml:3821]: This makes me think: let's test allocations inside multiple enclosing functions to demonstrate that all locks are walked, not just the closest.

File: typing/typecore.ml [hunk 2]
```diff
@@ -434,6 +434,10 @@ type expected_mode =
     field and [mode]: this field being [true] while [mode] being [global] is
     sensible, but not very useful as it will fail all expressions. *)
 
+    strictly_stack : bool;
```
Comment (dkalinichenko-js) [inline typing/typecore.ml:449]: Could this be unified with `strictly_local`? The comment for `strictly_local` after all mentions that it should force any allocations to be on the stack. I guess there is a difference with `local_` allowing things that can't be on the stack like modules while `stack_` forbids them?

File: typing/typecore.ml [hunk 3]
```diff
@@ -561,7 +566,8 @@ let mode_morph f expected_mode =
   let mode = as_single_mode expected_mode in
   let mode = f mode |> Mode.Value.disallow_left in
   let tuple_modes = None in
-  {expected_mode with mode; tuple_modes}
+  { expected_mode with mode; tuple_modes;
+    strictly_stack = false (* strictly_stack does not affect children *) }
```
Comment (dkalinichenko-js) [inline typing/typecore.ml]: Is this the right function to put `strictly_stack = here`? Is `mode_morph` only used for subexpressions?

File: typing/typecore.ml [hunk 4]
```diff
@@ -4649,16 +4666,32 @@ let collect_unknown_apply_args env funct ty_fun mode_fun rev_args sargs ret_tvar
...
+      let ty_fun_is_arrow =
+        match get_desc ty_fun' with
+        | Tarrow _ -> true
+        | _ -> false
+      in
+      (* CR shsong: Alternative design: only register_allocation_mode if
+          partial_app = is_partial_apply untyped_args (where untyped_args
+          are in the return value) is false, since if partial_app is true,
+          there are Omitted args and the code walks the locks already *)
+      if ty_fun_is_arrow then
```
Comment (dkalinichenko-js) [inline typing/typecore.ml:4960]: I don't think this is the right place to put in this check. Ideally, we should put it closest to the place where the allocation happens. Also, why isn't it calling `register_allocation_mode`?

Btw, this part of the compiler changed during the 5.4 merge, so looks like you'll need to rebase it anyway.

File: typing/typecore.ml [hunk 5]
```diff
@@ -7809,6 +7843,10 @@ and type_expect_
       let to_unify = Predef.type_lazy_t ty in
       with_explanation (fun () ->
         unify_exp_types loc env to_unify (generic_instance ty_expected));
+      (* Allocation axis check: constructing a lazy block allocates, but
```
Comment (dkalinichenko-js) [inline typing/typecore.ml]: Should we call `register_allocation_mode` here and force it to be global?

File: typing/typecore.ml [hunk 6]
```diff
@@ -7821,6 +7859,10 @@ and type_expect_
   | Pexp_object s ->
       Env.check_no_open_quotations loc env Object_qt;
       submode ~loc ~env Value.legacy expected_mode;
+      (* Allocation axis check: constructing an object block allocates,
+         but register_allocation_mode is not called, so we manually
+         walk locks here *)
+      Env.walk_locks_for_allocation ~env (loc, Hint.Allocation);
```
Comment (dkalinichenko-js) [inline typing/typecore.ml]: Same comment.

File: typing/typecore.ml [hunk 7]
```diff
@@ -11529,6 +11582,10 @@ and type_comprehension_expr ~loc ~env ~ty_expected ~attributes cexpr =
        "What modes should comprehensions use?", above *)
     type_expect new_env mode_legacy sbody (mk_expected element_ty)
   in
+  (* Allocation axis check: comprehension expr allocates, but
```
Comment (dkalinichenko-js) [inline typing/typecore.ml]: Same comment.

**Review by dkalinichenko-js — COMMENTED** [review]
Some comments.

File: typing/typecore.ml (hunk 2 above)
Comment (zhizhuoxin) [reply typing/typecore.ml:449]: Some uncaught exception arises when trying to merge the two variables naively. A CR is left in the code. We can review this later.

File: typing/typecore.ml (hunk 3 above)
Comment (zhizhuoxin) [reply typing/typecore.ml]: You are right. We do not need this. I removed this and add tests to indicate why we should not add this line.

File: typing/typecore.ml (hunk 4 above)
Comment (zhizhuoxin) [reply typing/typecore.ml:4960]: This check for partial application is not sufficient.
(According to our discussion), we will fix it in the next PR when we fix issues for currying (CR is left in the code).

register_allocation_mode is used (instead of directly calling walk locks) (for this and following similar cases).

File: typing/env.ml (hunk 1 above)
Comment (zhizhuoxin) [reply typing/env.ml:3821]: Test 5b in testsuite/tests/typing-modes/zero_alloc.ml is added to test this.

File: typing/env.ml [hunk 8]
```diff
@@ -3814,6 +3814,21 @@ let walk_locks_for_legacy_construct ~env pp =
        (Mode.Value.disallow_right Mode.Value.legacy) None locks
       : Mode.Value.l)
 
+(** Registers a use of an allocation, constraining every enclosing closure lock.
+    Used for constructs with allocations that force enclosing functions to be
+    alloc (rather than noalloc_strict) *)
+(* CR shsong: currently it only considers noalloc_strict and alloc,
+    need to customize this to support noalloc later *)
+let walk_locks_for_allocation ~env pp =
```
Comment (dkalinichenko-js) [inline typing/env.ml:3826]: We can share most of the code with `walk_locks_for_legacy_construct`, changing just the mode.

File: testsuite/tests/typing-modes/zero_alloc.ml [hunk 9]
```diff
@@ -382,8 +382,1290 @@ Line 1, characters 66-67:
...
+12 |   exclave_ stack_ (lazy (Just a))
+                       ^^^^^^^^^^^^^^^
+Error: The allocation is "alloc"
+       but is expected to be "noalloc_strict"
+         because it is used inside the function at lines 11-12, characters 40-33
+         which is expected to be "noalloc_strict".
+|}]
+
+(* a new allocation captured as an argument of a partial application *)
+let (stack_inner_partial @ noalloc_strict)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/zero_alloc.ml]: I believe this fails due to partial application in the function and not due to . Can you add a test that's just a multi-argument function with a blank body?

File: testsuite/tests/typing-modes/zero_alloc.ml [hunk 10]
```diff
@@ -382,8 +382,1290 @@ Line 1, characters 66-67:
...
+   lock-walk through [register_allocation_mode]. The others turned out NOT to be
+   problems and are kept for contrast. *)
+
+(* CR shsong (false positive): array indexing uses the [%array_safe_get]
+   primitive, whose result is [@local_opt] (Prim_poly). It does NOT allocate, but
+   the Prim_poly registration in [type_ident] goes through
+   [register_allocation_mode], which now walks locks, so it is wrongly rejected
+   -- note the error is "The allocation is alloc". It should be accepted once the
+   Prim_poly site is exempted from walking. *)
+let (array_get_prim_poly @ noalloc_strict) (a : int array) (i : int) = a.(i)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/zero_alloc.ml]: I believe this also fails due to multi-argument function issues.

File: testsuite/tests/typing-modes/zero_alloc.ml [hunk 11]
```diff
@@ -473,12 +1758,29 @@ Error: The module is "alloc"
...
+   created at [noalloc_strict]: the binding annotation requires the object value
+   to be [noalloc_strict], but its construction forces it to [alloc]. *)
+let (o @ noalloc_strict) = object method get = 0 end
+[%%expect{|
+Line 1, characters 27-52:
+1 | let (o @ noalloc_strict) = object method get = 0 end
+                               ^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This value is "alloc" but is expected to be "noalloc_strict".
+|}]
+
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/zero_alloc.ml:1840]: Stale comment?

File: testsuite/tests/typing-modes/zero_alloc.ml [hunk 12]
```diff
@@ -489,7 +1791,13 @@ let (secretly_allocates' @ noalloc) () = exclave_
...
+Line 3, characters 30-60:
+3 |  (escape_hatch <- stack_ Some (fun () -> { x = 1.; y = 2. })) [@zero_alloc];
+                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The allocation is "alloc"
+       but is expected to be "noalloc"
+         because it is used inside the function at lines 1-6, characters 36-17
+         which is expected to be "noalloc".
 |}]
 
 (* CR-soon shsong: error expected -- cannot call [f ()] because it is alloc *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/zero_alloc.ml]: Also stale comment?

File: testsuite/tests/typing-modes/zero_alloc.ml [hunk 13]
```diff
@@ -473,12 +1758,29 @@ Error: The module is "alloc"
...
+Line 2, characters 30-58:
+2 |   let (whitewashed @ alloc) = fun () -> { x = 1.; y = 2. } in
+                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The allocation is "alloc"
+       but is expected to be "noalloc"
+         because it is used inside the function at lines 1-3, characters 35-16
+         which is expected to be "noalloc".
 |}]
 
 (* CR-soon shsong: error expected -- cannot assign alloc func to escape_hatch ref *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/zero_alloc.ml]: Also stale comment?

File: testsuite/tests/typing-modes/zero_alloc.ml (hunk 9 above)
Comment (dkalinichenko-js) [reply testsuite/tests/typing-modes/zero_alloc.ml]: Feel free to resolve this since it'll be tested in a subsequent PR.

File: testsuite/tests/typing-modes/zero_alloc.ml [hunk 14]
```diff
@@ -382,8 +382,1290 @@ Line 1, characters 66-67:
...
+    introduce their own allocation, so the [stack_] exemption must NOT reach
+    them. It does not: [stack_] only marks its top operand, and every inner
+    allocation below is registered with [strictly_stack = false] -- either
+    because its mode is built by [mode_default] (constructor args, tuple
+    components, record fields, function bodies) or because [mode_morph] resets
+    [strictly_stack] for children (lazy thunk body, partial-application args). *)
+
+(* lazy thunk body holds a new allocation *)
+let (stack_inner_lazy @ noalloc_strict) (a : int) =
+  exclave_ stack_ (lazy (Just a))
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/zero_alloc.ml]: I think `lazy` itself allocates, so the `Just a` is moot.

File: testsuite/tests/typing-modes/zero_alloc.ml [hunk 15]
```diff
@@ -382,8 +382,1290 @@ Line 1, characters 66-67:
...
+                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The allocation is "alloc"
+       but is expected to be "noalloc_strict"
+         because it is used inside the function at line 1, characters 36-82
+         which is expected to be "noalloc_strict".
+|}]
+
+(* An object with no allocating method (here only a value field) still
+   allocates the object block itself, which the [Pexp_object] walk detects,
+   so it is rejected even though [get] does not allocate. *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/zero_alloc.ml]: Stale comment about `get`?

**Review by dkalinichenko-js — COMMENTED** [review]
Looks mostly good, minor comments.

File: testsuite/tests/typing-modes/zero_alloc.ml (hunk 10 above)
Comment (zhizhuoxin) [reply testsuite/tests/typing-modes/zero_alloc.ml]: Modified the test to suppress the multi-argument function issue so that it tests what we want to see.

File: testsuite/tests/typing-modes/zero_alloc.ml (hunk 14 above)
Comment (zhizhuoxin) [reply testsuite/tests/typing-modes/zero_alloc.ml]: Removed this to avoid confusion since we already have another test for lazy.

**Review by dkalinichenko-js — APPROVED** [review]
